proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,161 @@
+* *2 - there are no quantised buckets configured for this metric. it gives the user count/sum/max values only. it would be
+  hard to use this metric to understand something about produce requests or fetch responses.
+* *1 / *2 - we have contradictory meanings for the `flowing` label.  in *2, requests flow `UPSTREAM`, responses flow `DOWNSTREAM`.
+  In *1 requests flow `DOWNSTREAM` which makes no sense.
+
+### General short-comings/weaknesses
+
+* The metrics endpoint populates `HELP` for each metric, but this is populated with nothing more than the metric's name.
+  We ought to be giving a clear description of the metric.
+* Some of the label names don't follow Prometheus naming conventions.  Prometheus recommends snake case whereas we have used camel case in some cases.
+
+## Use Cases
+
+In general, metrics exist to help a user understand the health and performance of the proxy. They help inform
+capacity planning, drive system alerts reporting abnormal conditions and help with problem-solving (things like
+diagnosing failures or system slow-downs).
+
+Introducing a proxy to a Kafka system will increase latency of the system. Users will need a mechanism to understand how
+much latency is being added for their particular use-cases and have the ability to monitor this latency over time.
+
+Kroxylicious Filters themselves have the potential to alter behaviour of the Kafka system.  Users will need insights into
+performance and other issues that might result from filter behaviour. Specifically:
+
+* filters have the ability to change the size of the RPCs as they traverse the proxy (record encryption)
+* filters have the ability to change the flow of messages within the Kafka system. Specifically, they can short-circuit
+  requests by producing their own responses (the broker never sees the request) and introduce their own requests.
+
+## Proposal
+
+### Deprecate `kroxylicious_inbound*`/`kroxylicious_payload_size_bytes` replace with new `kroxylicious_(downstream|upstream)_kafka_rpc_message_size_bytes` metrics
+
+This proposal suggests that we deprecate the `kroxylicious_inbound*`/`kroxylicious_payload_size_bytes` metrics and replace
+with two new **distribution** (aka histogram) metrics `kroxylicious_downstream_kafka_rpc_message_size_bytes` and `kroxylicious_upstream_kafka_rpc_message_size_bytes`.
+
+The new "message_size" differ from the old "payload_size" in several key ways:
+1. New "message_size" will be used to track both **opaque** and **decoded** messages ("payload_size" tracked only decoded messages).
+2. Downstream "message size" will track [message size](https://kafka.apache.org/protocol#:~:text=DESCRIPTION-,message_size,-The%20message_size%20field)
+   as messages come in from or leave for the client.  Upstream "message size" will track message size as messages go to or arrive from the broker.
+3. The design of the metric allows it to be used to track messages that originate within a filter.
```

Comment:
Just to make sure I'm understanding this point. The metric will allow us to count all messages being sent to the broker by the proxy, be they a message from a client or invented by a filter.

Rather than allow us to identify which filter the message came from.

Reply:
Yes - exactly